City-of-Helsinki/helsinki-paatokset
# PP-221: Callback functionality and single item migrations

**To test**
* Checkout branch and run `make new`
* Connect to VPN and get a token from the Ahjo Open ID module: https://helsinki-paatokset.docker.so/fi/admin/ahjo-open-id
* SSH into the container and run: `drush ap:fs`
* Check that the single proxy endpoints work. Each one should output some content.
  * Meetings: https://helsinki-paatokset.docker.so/fi/ahjo-proxy/meetings/single/U0298020221
  * Cases: https://helsinki-paatokset.docker.so/fi/ahjo-proxy/cases/single/HEL-2019-011500
  * Decisions: https://helsinki-paatokset.docker.so/fi/ahjo-proxy/decisions/single/%7B520175BB-B575-426D-8561-FE63BC591402%7D
  * Trustees: https://helsinki-paatokset.docker.so/ahjo-proxy/trustees/single/arhinpa
* Run single migrations via drush:
  * `drush ahjo-proxy:update-entity meetings U0298020221 -v`
  * `drush ahjo-proxy:update-entity cases HEL-2019-011500 -v`
  * `drush ahjo-proxy:update-entity decisions {520175BB-B575-426D-8561-FE63BC591402} -v`
  * `drush ahjo-proxy:update-entity trustees arhinpa -v`
  * An invalid ID should result in the migration completing with status 0, not 1: `drush ahjo-proxy:update-entity meetings ABCDEDFG -v`
* Check the content listing to make sure those entities were actually created: https://helsinki-paatokset.docker.so/fi/admin/content/
  * Check that the migration actually updates content by changing one of the titles and running the command again. The title you added should get overridden
* Visit this URL a few times: https://helsinki-paatokset.docker.so/ahjo_api/subscriber/test
* Send POST requests to the same URL with this as the body (raw JSON, not form-data or x-www-form-urlencoded):
```
{
    "updatetype": "Updated",
    "id": "{22A36C7B-4CC5-4CCD-8B21-5AF0D45B481C}"
}
```
```
{
    "updatetype":Added",
    "id": "U12345"
}
```
* Run this command: `drush ac:l`. It should list the callbacks you just made
* Run the queue with: `drush queue:run ahjo_api_subscriber_queue`. (This will result in errors, which is fine)
  * Check the database log: https://helsinki-paatokset.docker.so/fi/admin/reports/dblog?type%5B%5D=ahjo_api_subscriber_queue
  * You should see status log about the first few requests being skipped and errors for the last two
  * Run `drush ac:l`. The queue should be empty.
  * Run `drush cron`, then `drush ac:l`. The callbacks that resulted in errors should be visible in the queue again.
  * Delete the items with `drush ahjo-callback:delete-item [id]` (replace the [id] with whatever the queue ID is).
* Run `drush ahjo-proxy:generate-queue -v`. This should add hundreds of items to the queue (these are copied over from the test site),
  * Run `drush ac:l` again. You should see all the hundreds of callbacks with about 30 unique IDs for decisions and meetings.
  * Run the queue again with: `drush queue:run ahjo_api_subscriber_queue`. This will take a long time (about 10 minutes) and will result in a few errors).
  * Check that more decisions and meetings were actually created:
    * https://helsinki-paatokset.docker.so/admin/content/decisions
    * https://helsinki-paatokset.docker.so/admin/content/meetings
